src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,899 @@
+    if(dfcc_is_loop_latch(target) && dfcc_has_loop_id(target, loop_id))
+    {
+      result = target;
+    }
+  }
+  return result;
+}
+
+static optionalt<goto_programt::targett> check_has_contract_rec(
+  const dfcc_loop_nesting_grapht &loop_nesting_graph,
+  const std::size_t loop_idx,
+  const bool must_have_contract)
+{
+  const auto &node = loop_nesting_graph[loop_idx];
+  if(must_have_contract && !has_contract(node.latch))
+    return node.head;
+
+  for(const auto pred_idx : loop_nesting_graph.get_predecessors(loop_idx))
+  {
+    auto result = check_has_contract_rec(
+      loop_nesting_graph, pred_idx, has_contract(node.latch));
+    if(result.has_value())
+      return result;
+  }
+  return {};
+}
+
+/// Traverses the loop nesting graph from top level loops and checks if all
+/// loops nested in loops that have contracts also have contracts.
+/// Return the head of the first offending loop if it exists, nothing otherwise.
+static optionalt<goto_programt::targett> check_inner_loops_have_contracts(
+  const dfcc_loop_nesting_grapht &loop_nesting_graph)
+{
+  for(std::size_t idx = 0; idx < loop_nesting_graph.size(); idx++)
+  {
+    if(loop_nesting_graph.get_successors(idx).empty())
+    {
+      auto result = check_has_contract_rec(loop_nesting_graph, idx, false);
+      if(result.has_value())
+        return result;
+    }
+  }
+  return {};
+}
+
+/// Tags instructions of loops found in \p loop_nesting_graph with the loop
+/// identifier of the innermost loop it belongs to, and the loop instruction
+/// type : head, body, exiting or latch.
+static void tag_loop_instructions(
+  goto_programt &goto_program,
+  dfcc_loop_nesting_grapht &loop_nesting_graph)
+{
+  for(const auto &idx : loop_nesting_graph.topsort())
+  {
+    auto &node = loop_nesting_graph[idx];
+    auto &head = node.head;
+    auto &latch = node.latch;
+    auto &instruction_iterators = node.instructions;
+
+    dfcc_set_loop_head(head);
+    dfcc_set_loop_latch(latch);
+
+    for(goto_programt::targett t : instruction_iterators)
+    {
+      // Skip instructions that are already tagged and belong to inner loops.
+      auto loop_id_opt = dfcc_get_loop_id(t);
+      if(loop_id_opt.has_value())
+        continue;
+
+      dfcc_set_loop_id(t, idx);
+
+      if(t != head && t != latch)
+        dfcc_set_loop_body(t);
+
+      if(t->is_goto() && !instruction_iterators.contains(t->get_target()))
+      {
+        dfcc_set_loop_exiting(t);
+      }
+    }
+  }
+
+  auto top_level_id = loop_nesting_graph.size();
+
+  // tag remaining instructions as top level
+  for(auto target = goto_program.instructions.begin();
+      target != goto_program.instructions.end();
+      target++)
+  {
+    // Skip instructions that are already tagged (belong to loops).
+    auto loop_id_opt = dfcc_get_loop_id(target);
+    if(loop_id_opt.has_value())
+    {
+      continue;
+    }
+    dfcc_set_loop_id(target, top_level_id);
+    dfcc_set_loop_top_level(target);
+  }
```

Comment:
I keep wondering: why do we actually need to tag these? Can't we just leave them untagged?

Reply:
the instrumentation passes that follow expect to find a tag on each instruction, and use it as a key to query the write set variable from the dfcc_cfg_info class. 

1) Tagging all instructions amounts to using  `size_t` as key type and a  special value to represent the top-level scope.
2) Not tagging some instructions amounts to using `optionalt<size_t>` as key type, with none the for top level scope.

I think 1) allows for better sanity checks.

With 1), if you discover an instruction without tag then it is an error for sure. With 2), it could be because it is a top level instruction, or it could be a loop instruction that was skipped because of a bug somewhere.